job/service.go

```diff
@@ -25,13 +28,33 @@ const (
 	ConcurrentLimit        = 600
 )
 
-var errDependencyResolution = fmt.Errorf("dependency resolution")
+var (
+	errDependencyResolution = fmt.Errorf("dependency resolution")
+
+	resolveDependencyFailureGauge = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "runtime_job_dependency_failure",
+		Help: "Number of job dependency resolution failure",
+	})
+	resolveDependencySuccessGauge = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "runtime_job_dependency_success",
+		Help: "Number of job dependency resolution succeed",
+	})
+	resolveDependencyHistogram = promauto.NewHistogram(prometheus.HistogramOpts{
+		Name: "runtime_job_dependency_histogram",
+		Help: "Duration of resolving job dependency",
+	})
+)
 
 type AssetCompiler func(jobSpec models.JobSpec, scheduledAt time.Time) (models.JobAssets, error)
 
 // DependencyResolver compiles static and runtime dependencies
 type DependencyResolver interface {
 	Resolve(ctx context.Context, projectSpec models.ProjectSpec, jobSpec models.JobSpec, observer progress.Observer) (models.JobSpec, error)
+	ResolveAndPersist(ctx context.Context, projectSpec models.ProjectSpec, jobSpec models.JobSpec, observer progress.Observer) error
```

Comment:
agree, I think it will reduce the complexity. however, in the existing resolve function, there is a `resolveHookDependencies` process, which is not needed (as we are going to store the requested and fetch all after). should we also decouple it from `Resolve` or keep it until we refactor job `Sync`?

Reply:
As discussed, reuse resolve and introduce `Persist` instead of `ResolveAndPersist`